### PR TITLE
[B-04408] New Publish Based on Neighborhood Delegate Broadcast

### PR DESCRIPTION
**Pull request by neonphog**

### CODE REVIEW

File: crates/kitsune_p2p/direct/src/types/kdhash.rs [hunk 1]
```diff
@@ -26,6 +26,12 @@ pub trait KdHashExt: Sized {
     /// convert from kitsune op hash
     fn from_kitsune_op_hash(op_hash: &KitsuneOpHash) -> Self;
 
+    /// convert to kitsune op hash
```
Comment (Connoropolous) [inline crates/kitsune_p2p/direct/src/types/kdhash.rs:29]: Looks like the comment is about an OpHash while this is a Basis?
Same below

**Review by Connoropolous — COMMENTED** [review]
Was just scanning and saw something small

File: crates/kitsune_p2p/direct/src/types/kdhash.rs (hunk 1 above)
Comment (neonphog) [reply crates/kitsune_p2p/direct/src/types/kdhash.rs:29]:
```suggestion
    /// convert to kitsune basis
```

File: crates/kitsune_p2p/direct/src/types/kdhash.rs [hunk 2]
```diff
@@ -26,6 +26,12 @@ pub trait KdHashExt: Sized {
...
+    fn to_kitsune_basis(&self) -> Arc<KitsuneBasis>;
+
+    /// convert from kitsune op hash
```
Comment (neonphog) [inline crates/kitsune_p2p/direct/src/types/kdhash.rs:32]:
```suggestion
    /// convert from kitsune basis
```

File: crates/holochain_p2p/src/spawn/actor.rs [hunk 3]
```diff
@@ -931,30 +931,28 @@ impl HolochainP2pHandler for HolochainP2pActor {
...
     ) -> HolochainP2pHandlerResult<()> {
+        use kitsune_p2p_types::KitsuneTimeout;
+
         let space = dna_hash.into_kitsune();
-        let from_agent = from_agent.into_kitsune();
         let basis = dht_hash.to_kitsune();
+        let timeout = match timeout_ms {
+            Some(ms) => KitsuneTimeout::from_millis(ms),
+            None => KitsuneTimeout::from_millis(30_000),
+        };
```
Comment (maackle) [inline crates/holochain_p2p/src/spawn/actor.rs:947]: FYI

```suggestion
        let timeout = KitsuneTimeout::from_millis(timeout_ms.unwrap_or(30_000));
```

Also can 30k be a const?

File: crates/kitsune_p2p/kitsune_p2p/src/types/wire.rs [hunk 4]
```diff
@@ -32,20 +32,31 @@ kitsune_p2p_types::write_codec_enum! {
...
+            mod_idx.3: u32,
+            mod_cnt.4: u32,
+            data.5: WireData,
         },
 
-        /// "Notify" response from the remote.
-        NotifyResp(0x21) {
+        /// Fire-and-forget broadcast message.
+        /// uses low-level notify, not request
+        Broadcast(0x23) {
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/types/wire.rs:50]: Doesn't seem like a "broadcast" if there's a single to_agent target for this message..

Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/types/wire.rs:50]: hehe, functionally it's a broadcast, the idea is it goes to everyone in the neighborhood... there's just no way to do a true broadcast on wan udp architecture... with good reason : )

File: crates/holochain_p2p/src/spawn/actor.rs (hunk 3 above)
Comment (neonphog) [reply crates/holochain_p2p/src/spawn/actor.rs]: switched to implicit timeout from tuning_params

**Review by maackle — CHANGES_REQUESTED** [review]
I mainly just want to see more documentation in general. For the scope of this PR, I pointed out the main parts I see lacking docs.

I'm personally not that good at understanding exactly what code does by reading all of it, and I benefit a lot from having the big picture explained to me without having to piece it together from the details. And also having some guidance for the tricky bits within functions. It's a basic courtesy to the reader that goes a long way.

**Comment by neonphog** [comment]
@maackle 

> It's a basic courtesy to the reader that goes a long way.

I always `*`try`*` to add comments where I think they're needed. But it's hard to tell when I'm in the headspace of code I've written. Feel free to say you need more comments in sections before continuing to review something if they are needed.

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor.rs [hunk 5]
```diff
@@ -32,6 +32,19 @@ ghost_actor::ghost_chan! {
         /// Register space event handler
         fn register_space_event_handler(recv: futures::channel::mpsc::Receiver<KitsuneP2pEvent>) -> ();
 
+        /// Incoming Delegate Broadcast
+        /// We are being requested to delegate a broadcast to our neighborhood
+        /// on behalf of an author. `mod_idx` / `mod_cnt` inform us which
+        /// neighbors we are responsible for.
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/spawn/actor.rs:38]:
```suggestion
        /// neighbors we are responsible for.
        /// (See comments in actual method impl for more detail.)
```

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs [hunk 6]
```diff
@@ -27,6 +27,16 @@ ghost_actor::ghost_chan! {
         /// see if an agent is locally joined
         fn is_agent_local(agent: Arc<KitsuneAgent>) -> bool;
 
+        /// Incoming Delegate Broadcast
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:30]: In retrospect I guess this is actually where that comment could have gone, and the other trait method comment could refer to this one, but I'm just glad that it exists somewhere now.

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs [hunk 7]
```diff
@@ -164,6 +175,87 @@ impl SpaceInternalHandler for Space {
...
+
+            let info_list = fut.await?;
+
+            // for all agents in the gathered list, check the modulo params
+            // i.e. if `agent.get_loc() % mod_cnt == mod_idx` we know we are
+            // responsible for delegating the broadcast to that agent.
+            let mut all = Vec::new();
+            for info in info_list
+                .into_iter()
+                .filter(|info| info.agent.get_loc() % mod_cnt == mod_idx)
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:222]: I'm totally cool with this now that there are comments, but just an idea for the future that could enhance readability even more (not asking for this):

`mod_cnt` and `mod_idx` could be encapsulated in a struct like `DelegateParams {mod_cnt, mod_idx}`, and then Agent could have a method like `should_delegate(delegate_params: DelegateParams) -> bool`. Then the essence of conditional delegation is encapulated and there's more of a single focal point (the struct) for that aspect of publishing.

**Review by maackle — APPROVED** [review]
Awesome, thank you, I like this level of detail.

I've had some built-up frustration about docs, not just from you, and some of it came out in this review. I get that you want to write docs but that it seems less compelling when you're already deep in the code. I also think you're especially good at reading and thinking in code. So maybe just imagine you're not : )

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs (hunk 6 above)
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:30]: coppied the comment here too :+1:

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs (hunk 7 above)
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:226]: I get where you're going from a code readability perspective... but from a serialization perspective, there's a tension in that the serialized api is effectively public.

The codec generation macro puts all the fields in a tuple so we don't waste space sending the field names over and over, and it's relatively simple at the moment, and doesn't support flattening.

**Pull request merged**